extensions/amp-flying-carpet/0.1/amp-flying-carpet.js

```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {CSS} from '../../../build/amp-flying-carpet-0.1.css';
+import {Layout} from '../../../src/layout';
+import {isExperimentOn} from '../../../src/experiments';
+import {dev, user} from '../../../src/log';
+
+/** @const */
+const EXPERIMENT = 'amp-flying-carpet';
+
+/** @const */
+const TAG = 'amp-flying-carpet';
+
+
+class AmpFlyingCarpet extends AMP.BaseElement {
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.FIXED_HEIGHT;
+  }
+
+  /** @override */
+  buildCallback() {
+    this.isExperimentOn_ = isExperimentOn(this.getWin(), EXPERIMENT);
+    if (!this.isExperimentOn_) {
+      dev.warn(TAG, `Experiment ${EXPERIMENT} disabled`);
+      return;
+    }
+  }
+
+  assertPosition() {
+    const layoutBox = this.element.getLayoutBox();
+    const viewport = this.getViewport();
+    const viewportHeight = viewport.getHeight();
+    const docHeight = viewport.getScrollHeight();
+    // Hmm, can the page height change and affect us?
+    user.assert(
+      layoutBox.top >= viewportHeight,
```

Comment:
It is an odd restriction, but it's because of the effect itself. See http://output.jsbin.com/focazo. Anything in the first viewport will never "reveal" the bottom of the content, and anything in the last viewport will never reveal the top of the content. We might be able to use some perspective and scaling trickery to make them work in any viewport but I haven't investigated that yet.